### PR TITLE
Adjust failure level appropriately when dynamic validation is in play.

### PR DESCRIPTION
**Pull request by michaelcfanning**
@eddynaka

### CODE REVIEW

**Review by eddynaka — APPROVED** [review]

**Comment by michaelcfanning** [comment]
>                         case Validation.ValidatorReturnedIllegalValue:

Interested in your thoughts here. The plugins can return an arbitrary failure level when they perform work. But, I think we should have some default handling based on certain dynamic validation return values. But we shouldn't override the default/plugin failure level in other cases. See if you agree I've made good decisions. :) Since this condition indicates a misbehaving plug-in, I've made it an error for now. Really, we should probably log a separate error notification and then process the result as is.

---
Refers to: Src/Sarif.PatternMatcher/SearchSkimmer.cs:290 in 5a0ac21. [](commit_id = 5a0ac2193b06774395610efbc97b65ec6791c1ec, deletion_comment = False)

File: Src/Sarif.PatternMatcher/SearchSkimmer.cs [hunk 1]
```diff
@@ -292,23 +292,27 @@ private void RunMatchExpression(FlexMatch binary64DecodedMatch, AnalyzeContext c
                             // The validator returned a bad value.
                             // TODO: we should log this condition
                             // and then continue processing.
+                            level = FailureLevel.Error;
                             break;
                         }
 
                         case Validation.Valid:
                         {
+                            level = FailureLevel.Error;
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher/SearchSkimmer.cs:301]: >level [](start = 28, length = 5)

If we conclusively validate something, it's important! So we make it an error.

File: Src/Sarif.PatternMatcher/SearchSkimmer.cs [hunk 2]
```diff
@@ -292,23 +292,27 @@ private void RunMatchExpression(FlexMatch binary64DecodedMatch, AnalyzeContext c
...
                             validationState = " which was determined to be valid";
                             break;
                         }
 
                         case Validation.Invalid:
                         {
+                            level = FailureLevel.Warning;
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher/SearchSkimmer.cs:308]: >level [](start = 28, length = 5)

If you conclusively marked something as invalid, that's a warning only, it shouldn't ever break a process (like an error will).

File: Src/Sarif.PatternMatcher/SearchSkimmer.cs (hunk 1 above)
Comment (eddynaka) [reply Src/Sarif.PatternMatcher/SearchSkimmer.cs:301]: yeah, we should think how we will log, so we can decide in the future based on data.

Comment (eddynaka) [reply Src/Sarif.PatternMatcher/SearchSkimmer.cs:301]: but, ur decision is perfect based on our knowledge so far

File: Src/Sarif.PatternMatcher/SearchSkimmer.cs [hunk 3]
```diff
@@ -292,23 +292,27 @@ private void RunMatchExpression(FlexMatch binary64DecodedMatch, AnalyzeContext c
...
                             validationState = " which was determined to be invalid";
                             break;
                         }
 
                         case Validation.InvalidForConsultedAuthorities:
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher/SearchSkimmer.cs:313]: >InvalidForConsultedAuthorities [](start = 40, length = 30)

We did some checking and it *looks* invalid, so, we'll set it to a warning. This assumes the user has configured the scan to check for the most critical authorities.

**Comment by michaelcfanning** [comment]
>                         case Validation.Unknown:

So, in all cases, when the answer is Unknown, we preserve the failure level set by the rule/plugin.

---
Refers to: Src/Sarif.PatternMatcher/SearchSkimmer.cs:320 in 5a0ac21. [](commit_id = 5a0ac2193b06774395610efbc97b65ec6791c1ec, deletion_comment = False)

**Pull request merged**